flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -3212,17 +3238,21 @@ class ArrayExprLowering {
       const Fortran::evaluate::Expr<Fortran::evaluate::SomeType> &e) {
     struct Filter : public Fortran::evaluate::AnyTraverse<
                         Filter, std::optional<llvm::SmallVector<mlir::Value>>> {
-      using Base = Fortran::evaluate::AnyTraverse<
-          Filter, std::optional<llvm::SmallVector<mlir::Value>>>;
+      using RT = std::optional<llvm::SmallVector<mlir::Value>>;
+      using Base = Fortran::evaluate::AnyTraverse<Filter, RT>;
       using Base::operator();
 
-      Filter(const llvm::SmallVector<mlir::Value> init)
-          : Base(*this), bounds(init) {}
+      Filter(ArrayExprLowering *const ael) : Base(*this), ael(ael) {}
 
-      std::optional<llvm::SmallVector<mlir::Value>>
-      operator()(const Fortran::evaluate::ArrayRef &ref) {
+      RT operator()(const Fortran::evaluate::ArrayRef &ref) {
         if ((ref.base().IsSymbol() || ref.base().Rank() == 0) &&
             ref.Rank() > 0 && !ref.subscript().empty()) {
+          auto baseTy = ael->raiseBaseType(ref.base());
+          auto ty = ref.base().IsSymbol()
+                        ? baseTy
+                        : baseTy.cast<fir::RecordType>().getType(
+                              toStringRef(ref.GetLastSymbol().name()));
+          auto bounds = ael->getShape(ty);
```

Comment:
Does this assume that `ty` shape is compile time constant ?

Reply:
Yes. This bag of worms is still a TODO.